### PR TITLE
feat(node): implement catchup stream handler

### PR DESCRIPTION
**Pull request by fbozic**
- Implement a catchup stream handler. 
- Cleanup unused types from the initial implementation.

In the following PR I will implement initiation of stream connection and request sending. 

Part of: https://github.com/calimero-network/core/issues/184

### CODE REVIEW

File: crates/node/src/lib.rs [hunk 1]
```diff
@@ -704,12 +707,111 @@ impl Node {
...
         Ok(outcome)
     }
-
     async fn handle_stream(
         &mut self,
-        _stream: calimero_network::stream::Stream,
+        mut stream: calimero_network::stream::Stream,
     ) -> eyre::Result<()> {
-        // TODO: implement catchup mechanism once storage is ready
+        let request = match stream.next().await {
```
Comment (miraclx) [inline crates/node/src/lib.rs]: use `let-else` to short-circuit early, which helps compress indentation

```rust
let Some(message) = stream.next().await else {
    eyre::bail!("Stream closed unexpectedly")
}

let request = match serde_json::from_slice(&message?.data)? {
    types::CatchupStreamMessage::Request(req) => req,
    message => {
        eyre::bail!("Unexpected message: {:?}", message)
    }
}
```

File: crates/node/src/lib.rs [hunk 2]
```diff
@@ -704,12 +707,111 @@ impl Node {
...
+            Some(message) => match message {
+                Ok(message) => match serde_json::from_slice(&message.data)? {
+                    types::CatchupStreamMessage::Request(req) => req,
+                    message => {
+                        eyre::bail!("Unexpected message: {:?}", message)
+                    }
+                },
+                Err(err) => eyre::bail!(err),
```
Comment (miraclx) [inline crates/node/src/lib.rs]: `?` would've helped here

File: crates/node/src/catchup.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    pub(crate) async fn flush_with_error(
+        &mut self,
+        error: &types::CatchupError,
+    ) -> eyre::Result<()> {
+        self.flush().await?;
+
+        let message = serde_json::to_vec(error)?;
+        self.stream
+            .send(calimero_network::stream::Message { data: message })
+            .await?;
```
Comment (miraclx) [inline crates/node/src/catchup.rs:68]: if we're not clearing the pending messages from the queue, why send the error separately?

File: crates/node/src/lib.rs [hunk 4]
```diff
@@ -704,12 +707,111 @@ impl Node {
...
+            }
+
+            key = calimero_store::key::ContextTransaction::new(
+                request.context_id,
+                transaction.prior_hash.into(),
+            );
+        }
+
+        if !last_executed_transaction_hash_discovered {
+            let message = serde_json::to_vec(&types::CatchupError::TransactionNotFound {
```
Comment (miraclx) [inline crates/node/src/lib.rs]: just noting that - depending on how the transactions are applied on receipt, we may need a different error for this case

`TransactionNotFound` can be quickly returned without traversing tx links

Just do a `get` on the last executed we're given so failure is instant and we don't spend a lot of time looking for it

This case, however, represents the potential for us to have transactions that aren't part of the canonical history

Which, should be managed such that it doesn't happen (but \*might\* warrant a separate error case)

File: crates/node/src/lib.rs [hunk 5]
```diff
@@ -704,12 +707,111 @@ impl Node {
...
+            batch_writer
+                .send(types::TransactionWithStatus {
+                    transaction_hash: hash.into(),
+                    transaction: calimero_primitives::transaction::Transaction {
+                        context_id: request.context_id,
+                        method: transaction.method.into(),
+                        payload: transaction.payload.into(),
+                        prior_hash: calimero_primitives::hash::Hash::from(transaction.prior_hash),
+                    },
+                    status: types::TransactionStatus::Executed,
```
Comment (miraclx) [inline crates/node/src/lib.rs:793]: Hm, is this useful?

Comment (fbozic) [reply crates/node/src/lib.rs:793]: I think yes. When received, executed transactions should be executed immediately and transactions from the pool should be inserted in the pool (and executed later once TransactionConfirmation is received). Let me know if I am missing something.

File: crates/node/src/catchup.rs (hunk 3 above)
Comment (fbozic) [reply crates/node/src/catchup.rs:68]: I don't understand the question. First I flush the batch and after that, I send an error message.

Comment (miraclx) [reply crates/node/src/catchup.rs:68]: yeah, why not add the error to the batch to be sent together?

Comment (miraclx) [reply crates/node/src/catchup.rs:68]: ah - different types, the batch isn't a "message" it's transactions, oke oke

File: crates/node/src/lib.rs (hunk 5 above)
Comment (miraclx) [reply crates/node/src/lib.rs:793]: Ah, yeah.. 👌🏽

File: crates/node/src/catchup.rs (hunk 3 above)
Comment (fbozic) [reply crates/node/src/catchup.rs:68]: yeah that is why. let me know if we want to open a issue to implement BatchMessage on the stream protocol itself.

Comment (miraclx) [reply crates/node/src/catchup.rs:68]: ye, it'll be useful for Blob API for example, but for now this is okay

**Review by miraclx — COMMENTED** [review]
lgtm so far, will you be implementing the apply side in this PR or a follow-up?

**Review by miraclx — APPROVED** [review]
lgtm

File: crates/node/src/catchup.rs (hunk 3 above)
Comment (fbozic) [reply crates/node/src/catchup.rs:68]: https://github.com/calimero-network/core/issues/449

**Pull request merged**